services/director-v2/src/simcore_service_director_v2/modules/redis.py

```diff
@@ -0,0 +1,180 @@
+
+DEFAULT_LOCKS_PER_NODE: Final[int] = 2
+EXTEND_TASK_ATTR_NAME = "extend_task"
+
+DockerNodeId = str
+
+
+logger = logging.getLogger(__name__)
+
+redis_retry_policy = dict(
+    wait=wait_random(5, 10),
+    before_sleep=before_sleep_log(logger, logging.WARNING),
+    reraise=True,
+)
+
+
+def setup(app: FastAPI):
+    @retry(**redis_retry_policy)
+    async def on_startup() -> None:
+        app.state.redis_lock_manager = await RedisLockManager.create(app)
+
+    async def on_shutdown() -> None:
+        lock_manager: RedisLockManager = app.state.redis_lock_manager
+        await lock_manager.close()
+
+    app.add_event_handler("startup", on_startup)
+    app.add_event_handler("shutdown", on_shutdown)
+
+
+@dataclass
+class LocksPerNodeProvider:
+    # NOTE: added here for future use, in this iteration just
+    # returns a constant
+
+    async def get(  # pylint: disable=unused-argument,no-self-use
+        self, docker_node_id: DockerNodeId
+    ) -> int:
+        return DEFAULT_LOCKS_PER_NODE
+
+
+@dataclass
+class RedisLockManager:
+    app: FastAPI
+    redis: Redis
+    lock_per_node_provider: LocksPerNodeProvider
+    lock_timeout: float = 10.0
+
+    @classmethod
+    async def create(cls, app: FastAPI) -> "RedisLockManager":
+        settings: RedisSettings = app.state.settings.REDIS
+        redis = Redis.from_url(settings.dsn_locks)
+        lock_per_node_provider = LocksPerNodeProvider()
+        return cls(app=app, redis=redis, lock_per_node_provider=lock_per_node_provider)
+
+    @classmethod
+    def instance(cls, app: FastAPI) -> "RedisLockManager":
+        if not hasattr(app.state, "redis_lock_manager"):
+            raise ConfigurationError(
+                "RedisLockManager client is not available. Please check the configuration."
+            )
+        return app.state.redis_lock_manager
+
+    @staticmethod
+    def _get_node_slots_key(docker_node_id: DockerNodeId) -> str:
+        return f"lock_manager.{docker_node_id}.lock_slots"
+
+    @staticmethod
+    def _get_node_lock_name(
+        docker_node_id: DockerNodeId, slot_index: NonNegativeInt
+    ) -> str:
+        return f"lock_manager.{docker_node_id}.lock_slot.{slot_index}"
+
+    async def get_node_slots(self, docker_node_id: DockerNodeId) -> int:
+        """get the total amount of slots available for the node"""
+        node_slots_key = self._get_node_slots_key(docker_node_id)
+        slots: Optional[bytes] = await self.redis.get(node_slots_key)
+        if slots is not None:
+            return int(slots)
+
+        default_slots = await self.lock_per_node_provider.get(docker_node_id)
+        await self.redis.set(node_slots_key, default_slots)
+        return default_slots
+
+    @cached_property
+    def lock_extend_interval(self) -> float:
+        return self.lock_timeout / 2
+
+    async def _extend_task(self, lock: Lock) -> None:
+        while True:
+            await asyncio.sleep(self.lock_extend_interval)
```

Comment:
Sorry, I do not totally understand this tuning to solve deadlocks. Moreover, what values do **guarantee** that deadlock will not exist?

Reply:
It could happen that the `director-v2` acquires all the locks and it is killed. At this point nothing will any longer be able to release them.
If the locks are not released, the system will no longer be able to save any data.

To allow for the locks to be released they must be acquire with a timeout and a task, which extends their timeout at regular intervals must be created.